### DISCUSSION TITLE
Keyboard version

### DISCUSSION THREAD

**Discussion by pokey** [cursorless-dev/cursorless#264] [Ideas]
Discussing keyboard version of cursorless

**Comment by anthonyjclark** [comment]
Hello, I watched your Strange Loop talk and I really like what I saw. This project is very inspiring. Thank you!

After reading the [modal keyboard interface](https://www.cursorless.org/docs/user/experimental/keyboard/modal/) documentation, I am interested in transitioning some of my current modal keyboard workflow to cursorless. I am wondering if you could provide some advise.

I currently use a custom modal keyboard configuration based on the [vscode-modal-editor](https://github.com/DCsunset/vscode-modal-editor#vscode-modal-editor) extension. I'd love to start gradually replace some of my custom functionality with cursorless.

For example, here is a snippet from my configuration that I use to select an argument/parameter:

```json5
		// ...
		"E": {
			// ...
			"a": [
				findText(false, false, true, true, "[\(,]"),
				{ command: "cursorRight", when: "_ctx.lineAt(_ctx.pos).text[_ctx.pos.character] == '('" },
				findText(true, true, false, true, "[\),]"),
				{ command: "cursorLeftSelect", when: "_ctx.lineAt(_ctx.pos).text[_ctx.pos.character] == ')'" },
			],
		// ...
```

I'd like to replace this with the tree-sitter commands in cursorless---with the idea to eventually replace most (if not all) of my current setup with cursorless (including jumping).

Here are two questions:

1. How do I encode "select argument" as a vscode command using cursorless commands.  (I was digging through the extension but I thought I might be better suited by asking you first. It looks like I would want to first execute a target command, and then an action.)
2. How do I disable some functionality (e.g., hats) while I decide what I'd like to use from cursorless? (I understand that this one might not be reasonable given the project's goals.)

**Reply by josharian** [reply]
Hi! I'm not @pokey but I'll share my 2c. :)

FYI, the cursorless keyboard interface is pretty skeletal at this point. There's recent renewed interest among several folks in polishing it up (@Kn0rk, myself, pokey), but that hasn't happened yet. Just fair warning that going may be a bit bumpy in the near term.

The Cursorless way to select an argument is to refer to it directly (e.g. `"take arg air"`), rather than scanning forward to find it from the current position. In practice, that probably requires hats. :)

I don't think any of us has really put much thought into how to teach keyboard Cursorless to someone who hasn't used spoken Cursorless, mostly because the keyboard version is still rudimentary. In the short term, you might be best off playing a little with spoken Cursorless and then migrating back from there.

Sorry that this isn't a very satisfying answer.

**Reply by anthonyjclark** [reply]
@josharian Thank you for your thoughts! (I'm open to anyone that wants to chat :)).

I like your idea of first coming at it from the spoken side, clearly the system was designed for voice. On the other hand, I might be of more use in writing documentation and tutorials if I come at it from the strictly keyboard side, since I won't have the "expert blindness" associated with knowing more about the spoken side.

If you all are open to documentation contributions, I might stay away from voice and struggle my way through the current documentation and implementation. Then I can document my process, ask some questions, and submit some new explanations.

Does that sounds reasonable?

**Reply by Kn0rk** [reply]
Thanks Josh for the mention I would not have noticed otherwise.

As mentioned above the keyboard implementation is still very experimental. However, if you are up for testing wonky keyboard setups that could change at any moment …  your feedback that would be invaluable. I think extensive documentation and tutorials only really make sense once there is a stable keyboard implementation though.

**Reply by josharian** [reply]
@anthonyjclark we are definitely acutely aware of the dangers of the curse of knowledge / expert blindness, and would very much appreciate your insights as you come up to speed. I just wanted to make sure you knew what you were getting yourself into. :)

We would be delighted to have documentation contributions, whether formal (PR) or informal (keeping good public notes as you go, here or elsewhere). Informal is probably better at least initially, unless it's super obvious to you what the right fix is, because as @Kn0rk notes, the implementation itself is likely to change substantially.

If you want a faster feedback cycle, there's an active community on the Talon Slack and the Hands-Free Discord. And there are thrice-a-week voice meetups. Happy to invite you to any/all of those. Whatever works best for you. Excited to have you here.

**Reply by anthonyjclark** [reply]
Thank you both. I'll keep it informal for now and add some notes to this discussion (hopefully somewhat organized by topic).

On a related note: do you all have a formal grammar for cursorless?

**Reply by josharian** [reply]
Sounds great.

Formal grammar: Sort of?

The actual grammar is written in .talon files, which are assembled by Talon into a DFA used during ASR. See e.g. https://github.com/cursorless-dev/cursorless/blob/main/cursorless-talon/src/cursorless.talon. You could probably reasonably turn that into a formal EBNF-style grammar, but to my knowledge, nobody has.

Note that even if you did so, almost all leaf nodes are hot-swappable, so that people can customize their spoken forms. So e.g. one person's "chuck branch ivy" might be equivalent to another's "clear case sit": clear/chuck both referring to the "remove" action, branch/case both referring to the "branch" scope, and ivy/sit both referring to the letter "i".

**Reply by josharian** [reply]
See https://github.com/cursorless-dev/cursorless/blob/main/cursorless-talon/src/spoken_forms.json for the default spoken forms.

**Reply by anthonyjclark** [reply]
Thanks again! I was creating a bit of a grammar in my head as I read through the docs and thought that an actual grammar might help me understand the model a bit better.

**Reply by pokey** [reply]
I'm planning to do a big push on keyboard stuff in the next week or so; getting lots of interest, so it's time. I like your idea of staying away from voice to keep a fresh eye. Stay tuned

**Reply by anthonyjclark** [reply]
After reading the documentation (and a few issues) and playing around with it a bit. Here are some early thoughts:

### Spoken Model

~~~text
command   : action target
target    : primitive | compound
primitive : mark | modifier+ mark?
compound  : range | list
~~~

### Keyboard Model

Enter listening mode with `ctrl-c`.

~~~text
command   : target action
target    : primitive | compound
primitive : mark | modifier+ mark?
compound  : range | list
~~~

### Thoughts

- I like the select-first model for keyboards
- The [modal keyboard interface](https://www.cursorless.org/docs/user/experimental/keyboard/modal/) videos are not overly useful standalone. I would guess you all are aware of this and that they are mostly placeholders.
- (Opinion incoming) I would introduce the concepts in a tutorial style prior to more thorough documentation. Something like this:

#### Tutorial Example

Let's say that you want to delete the `testing()` function below.

```javascript
function whatever() {|
    const hello = "world";
}

function testing() {
    const bar = "baz";
}
```

In Cursorless you will execute a *command* of this form: `command : target action`.

For our first example, the `target` is the entire `testing()` function (along with handling its white-space appropriately), and the `action` is deletion.

To initiate the command, you need to put your editor into a mode in which Cursorless can listen for input. In VSCode you can add the following to your `keybindings.json` file:

~~~json
    {
        "key": "ctrl+c",
        "command": "cursorless.keyboard.modal.modeOn",
        "when": "editorTextFocus"
    },
    {
        "key": "escape",
        "command": "cursorless.keyboard.escape",
        "when": "cursorless.keyboard.listening && editorTextFocus && !suggestWidgetMultipleSuggestions && !suggestWidgetVisible"
    },
~~~

Now you can type `ctrl-c` to type Cursorless commands, and `esc` to go back to typical VSCode behavior.

At this point, Cursorless does not know what keys it should associate with its various targets and actions. Let's add a few bindings.

~~~json
  "cursorless.experimental.keyboard.modal.keybindings.scopes": {
    "f": "namedFunction",
  },
  "cursorless.experimental.keyboard.modal.keybindings.actions": {
  },
  "cursorless.experimental.keyboard.modal.keybindings.colors": {
    "d": "default",
  },
~~~

You might have noticed that we've left the set of actions empty. That is because we need to handle some keys differently. `backspace` is one such key. It is natural to delete something with `backspace`, so let's add that to `keybindings.json`:

~~~json5
    // ...
    {
        "key": "backspace",
        "command": "cursorless.keyboard.targeted.runActionOnTarget",
        "args": "remove",
        "when": "cursorless.keyboard.modal.mode && editorTextFocus"
    }
    // ...
~~~

Now you can target any "default" mark (we'll explain marks soon) in `testing()`. (Assume we have a gray dot above the letter `z` in the token `baz`). To delete `testing()` we can now type: `ctrl-c` `dzf` `backspace`.

### Some thoughts on my tutorial

- It would be nice to use selectable text, but I don't know if that is possible with the hats.
- I imagine Cursorless will come with default bindings? If so, the language will need to change.
- I would continue the tutorial to build up a few scopes, a few actions, and a few colors.

**Reply by pokey** [reply]
Yeah I think tutorial is a great way to introduce information. We've been big on [diataxis](https://diataxis.fr/) philosophy lately. Re adding customizations, I do wonder if we really want to force users to build up their config from scratch, so I'm not sure that's where we'd want tutorial to start. Prob better to start with a full config and tutorial teaches them how to use it and maybe make small tweaks

**Reply by pokey** [reply]
Fwiw Cursorless keyboard is now expressed using [a formal grammar](https://github.com/cursorless-dev/cursorless/blob/main/packages/cursorless-vscode/src/keyboard/grammar/grammar.ne). It's still highly experimental, but is a nice compact way to express key sequences, and allows us to generate [nice railroads](https://www.cursorless.org/keyboard-modal-railroad) 😊

**Comment by pokey** [comment]
I wonder if we want to introduce some redundancy in the keyboard encoding. Cost of a typo seems higher than normal typing, and even there there is arguably the redundancy built into language. Cc/ @josharian

**Reply by josharian** [reply]
Maybe? I guess I'd be interested in the following redundant-ish additions:

* Press key to "submit" an action, rather than having it happen automatically. Provides an opportunity to fix typos and/or bail.
* Explicit keystroke to re-assign hats (https://github.com/cursorless-dev/cursorless/discussions/2111#discussioncomment-7935137)
* Added structure, e.g. all actions require "a" prefix.

(Should I file issues for these? Discussion items on #2111? Not sure how we want to keep this laundry list organized.)

**Reply by pokey** [reply]
I think pressing key to submit could make sense, tho I'm imagining getting a head of steam and moving fast enough that you wouldn't want to pause and check before you press that key. I guess the visual feedback from flashes as things get deleted / added helps, and then you can undo if something doesn't happen as you expect

I think we can just discuss these things in this thread, as they're all related to error tolerance

**Reply by Kn0rk** [reply]
I think i would also favor the undo action over an submit to excecute mode.
